canonicaljson.py

```diff
@@ -27,22 +32,37 @@
 __version__ = "1.6.0"
 
 
-def _default(obj):  # pragma: no cover
+def _default(obj: object) -> object:  # pragma: no cover
     if type(obj) is frozendict_type:
         # If frozendict is available and used, cast `obj` into a dict
-        return dict(obj)
+        return dict(obj)  # type: ignore[call-overload]
     raise TypeError(
         "Object of type %s is not JSON serializable" % obj.__class__.__name__
     )
 
 
+class Encoder(Protocol):  # pragma: no cover
+    def encode(self, data: object) -> str:
+        pass
+
+    def iterencode(self, data: object) -> Generator[str, None, None]:
+        pass
+
+    def __call__(self, *args: Any, **kwargs: Any) -> "Encoder":
+        pass
+
+
+class JsonLibrary(Protocol):
+    JSONEncoder: Encoder
+
+
 # Declare these in the module scope, but they get configured in
 # set_json_library.
-_canonical_encoder = None
-_pretty_encoder = None
+_canonical_encoder: Encoder = None  # type: ignore[assignment]
```

Comment:
Could you do something silly like `import json as _json` and then set these types to `_json.JSONEncoder`?

Reply:
Maybe---but that might cause further pain having to wrangle with `simplejson`.

(I was trying to avoid touching any of the import stuff---feels a bit icky.)

Reply:
Being completely honest a lot of this is overkill: the main benefit to Synapse is the existence of `-> bytes` and `-> Generator[bytes, None, None]` return types.